flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/TaskManagerPodDecorator.java

```diff
@@ -0,0 +1,136 @@
+import io.fabric8.kubernetes.api.model.ResourceRequirementsBuilder;
+import io.fabric8.kubernetes.api.model.Volume;
+import io.fabric8.kubernetes.api.model.VolumeMountBuilder;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.configuration.GlobalConfiguration.FLINK_CONF_FILENAME;
+import static org.apache.flink.kubernetes.utils.Constants.CONFIG_FILE_LOG4J_NAME;
+import static org.apache.flink.kubernetes.utils.Constants.CONFIG_FILE_LOGBACK_NAME;
+import static org.apache.flink.kubernetes.utils.Constants.CONFIG_MAP_PREFIX;
+import static org.apache.flink.kubernetes.utils.Constants.FLINK_CONF_VOLUME;
+
+/**
+ * Task manager specific pod configuration.
+ * */
+public class TaskManagerPodDecorator extends Decorator<Pod, FlinkPod> {
+
+	private static final String CONTAINER_NAME = "flink-task-manager";
+
+	private final TaskManagerPodParameter parameter;
+
+	public TaskManagerPodDecorator(TaskManagerPodParameter parameters) {
+		Preconditions.checkNotNull(parameters);
+		this.parameter = parameters;
+	}
+
+	@Override
+	protected Pod doDecorate(Pod pod, Configuration flinkConfig) {
+
+		String clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+		Preconditions.checkNotNull(clusterId, "ClusterId must be specified!");
+
+		Map<String, String> labels = org.apache.flink.kubernetes.kubeclient.decorators.LabelBuilder
+			.withExist(pod.getMetadata().getLabels())
+			.withTaskManagerComponent()
+			.toLabels();
+
+		pod.getMetadata().setLabels(labels);
+		pod.getMetadata().setName(this.parameter.getPodName());
+
+		Volume configMapVolume = new Volume();
+		configMapVolume.setName(FLINK_CONF_VOLUME);
+		configMapVolume.setConfigMap(new ConfigMapVolumeSourceBuilder()
+			.withName(CONFIG_MAP_PREFIX + clusterId)
+			.withItems(Arrays.asList(
+				new KeyToPath(FLINK_CONF_FILENAME, null, FLINK_CONF_FILENAME),
+				new KeyToPath(CONFIG_FILE_LOG4J_NAME, null, CONFIG_FILE_LOG4J_NAME),
+				new KeyToPath(CONFIG_FILE_LOGBACK_NAME, null, CONFIG_FILE_LOGBACK_NAME)))
+			.build());
+
+		pod.setSpec(new PodSpecBuilder()
+			.withVolumes(configMapVolume)
+			.withContainers(createTaskManagerContainer(flinkConfig))
+			.build());
+		return pod;
+	}
+
+	private Container createTaskManagerContainer(Configuration flinkConfig) {
+		Quantity taskManagerCpuQuantity = new Quantity(String.valueOf(parameter.getTaskManagerCpus()));
+		Quantity taskManagerMemoryQuantity = new Quantity(String.valueOf(
+			parameter.getTaskManagerMemory() + Constants.RESOURCE_UNIT_MB));
+		String image = flinkConfig.getString(KubernetesConfigOptions.CONTAINER_IMAGE);
+		String pullPolicy = flinkConfig.getString(KubernetesConfigOptions.CONTAINER_IMAGE_PULL_POLICY);
+		String flinkConfDirInPod = flinkConfig.getString(KubernetesConfigOptions.FLINK_CONF_DIR);
+		return new ContainerBuilder()
+			.withName(CONTAINER_NAME)
+			.withArgs(this.parameter.getArgs())
```

Comment:
seem quite similiar to FlinkMasterDeploymentDecorator ? could these codes be refined ?

Reply:
I will move the common logics to a utils method.